packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp

```diff
@@ -1184,137 +1462,147 @@ class FastILUPrec
             aRowMap (aRowMap_),
             aColIdx (aColIdx_),
             aRowIdx (aRowIdx_),
-            iperm (perm_)
+            iperm (perm_),
+            blockCrsSize(blockCrsSize_)
             {}
 
             // just calling CopySortedValsTag, or GetUpperTag
             FastILUPrec_Functor(ScalarArray aValIn_, OrdinalArray aRowMapIn_, OrdinalArray aColIdxIn_,
-                                ScalarArray aVal_, OrdinalArray aRowMap_, OrdinalArray aColIdx_) :
+                                ScalarArray aVal_, OrdinalArray aRowMap_, OrdinalArray aColIdx_, Ordinal blockCrsSize_) :
             aValIn (aValIn_),
             aRowMapIn (aRowMapIn_),
             aColIdxIn (aColIdxIn_),
             aVal (aVal_),
             aRowMap (aRowMap_),
-            aColIdx (aColIdx_)
+            aColIdx (aColIdx_),
+            blockCrsSize(blockCrsSize_)
             {}
 
             // just calling GetLowerTag
             FastILUPrec_Functor(ScalarArray aVal_, OrdinalArray aRowMap_, OrdinalArray aColIdx_,
                                 ScalarArray lVal_, OrdinalArray lRowMap_, OrdinalArray lColIdx_,
-                                ScalarArray diagElems_) :
+                                ScalarArray diagElems_, Ordinal blockCrsSize_) :
             aVal (aVal_),
             diagElems (diagElems_),
             aRowMap (aRowMap_),
             aColIdx (aColIdx_),
             lVal (lVal_),
             lRowMap (lRowMap_),
-            lColIdx (lColIdx_)
+            lColIdx (lColIdx_),
+            blockCrsSize(blockCrsSize_)
             {}
 
             // just calling SwapDiagTag
             FastILUPrec_Functor(const SwapDiagTag&, ScalarArray  lVal_, OrdinalArray  lRowMap_, OrdinalArray lColIdx_,
                                 ScalarArray utVal_, OrdinalArray utRowMap_, OrdinalArray utColIdx_,
-                                ScalarArray diagElems_) :
+                                ScalarArray diagElems_, Ordinal blockCrsSize_) :
             diagElems (diagElems_),
             lVal (lVal_),
             lRowMap (lRowMap_),
             lColIdx (lColIdx_),
             utVal (utVal_),
             utRowMap (utRowMap_),
-            utColIdx (utColIdx_)
+            utColIdx (utColIdx_),
+            blockCrsSize(blockCrsSize_)
             {}
 
             // just calling ColPerm
             FastILUPrec_Functor(OrdinalArray aColIdx_, OrdinalArray iperm_) :
             aColIdx (aColIdx_),
-            iperm (iperm_)
+            iperm (iperm_),
+            blockCrsSize(0)
             {}
-            
+
 
             // ------------------------------------------------
             // functor to load values
             // both matrices are sorted and, "a" (with fills) contains "aIn" (original)
             KOKKOS_INLINE_FUNCTION
             void operator()(const CopySortedValsTag &, const int i) const {
-                Ordinal aPtr = aRowMapIn[i];
+                Ordinal aPtr    = aRowMapIn[i];
```

Comment:
There were bugs in the original impl where it was performing operations on the wrong row. The aPtrEnd check will prevent those and I added that check to several functors.